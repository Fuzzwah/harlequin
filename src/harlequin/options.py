from __future__ import annotations

from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any, Callable

import click


class AbstractOption(ABC):
    def __init__(
        self,
        name: str,
        description: str,
        *args: Any,
        label: str | None = None,
        short_decls: list[str] | None = None,
        **kwargs: Any,
    ) -> None:
        if short_decls is None:
            short_decls = []
        self.name = name
        self.description = description
        self.label = label
        self.short_decls = short_decls

    @abstractmethod
    def to_click(self) -> Callable[[click.Command], click.Command]:
        pass


class TextOption(AbstractOption):
    def __init__(
        self,
        name: str,
        description: str,
        label: str | None = None,
        short_decls: list[str] | None = None,
        validator: Callable[[str], str] | None = None,
    ) -> None:
        super().__init__(name, description, label=label, short_decls=short_decls)
        self.validator = validator

    def to_click(self) -> Callable[[click.Command], click.Command]:
        def click_callback(
            ctx: click.Context, param: click.ParamType, value: str
        ) -> str:
            if self.validator is not None:
                try:
                    value = self.validator(value)
                except Exception as e:
                    raise click.BadParameter(str(e)) from e
            return value

        return click.option(
            f"--{self.name}",
            *self.short_decls,
            help=self.description,
            callback=click_callback,
        )


class ListOption(AbstractOption):
    def __init__(
        self,
        name: str,
        description: str,
        label: str | None = None,
        short_decls: list[str] | None = None,
    ) -> None:
        super().__init__(name, description, label=label, short_decls=short_decls)

    def to_click(self) -> Callable[[click.Command], click.Command]:
        return click.option(
            f"--{self.name}",
            *self.short_decls,
            help=self.description,
            multiple=True,
        )


class PathOption(AbstractOption):
    def __init__(
        self,
        name: str,
        description: str,
        label: str | None = None,
        short_decls: list[str] | None = None,
        exists: bool = False,
        file_okay: bool = True,
        dir_okay: bool = True,
        resolve_path: bool = False,
        path_type: type | None = Path,
    ) -> None:
        super().__init__(name, description, label=label, short_decls=short_decls)
        self.exists = exists
        self.file_okay = file_okay
        self.dir_okay = dir_okay
        self.resolve_path = resolve_path
        self.path_type = path_type

    def to_click(self) -> Callable[[click.Command], click.Command]:
        return click.option(
            f"--{self.name}",
            *self.short_decls,
            help=self.description,
            type=click.Path(
                exists=self.exists,
                file_okay=self.file_okay,
                dir_okay=self.dir_okay,
                resolve_path=self.resolve_path,
                path_type=self.path_type,
            ),
        )


class SelectOption(AbstractOption):
    def __init__(
        self,
        name: str,
        description: str,
        choices: list[str],
        label: str | None = None,
        short_decls: list[str] | None = None,
    ) -> None:
        super().__init__(name, description, label=label, short_decls=short_decls)
        self.choices = choices

    def to_click(self) -> Callable[[click.Command], click.Command]:
        return click.option(
            f"--{self.name}",
            *self.short_decls,
            help=self.description,
            type=click.Choice(choices=self.choices, case_sensitive=False),
        )


class FlagOption(AbstractOption):
    def to_click(self) -> Callable[[click.Command], click.Command]:
        return click.option(
            f"--{self.name}", *self.short_decls, help=self.description, is_flag=True
        )


HarlequinAdapterOption = AbstractOption
HarlequinCopyOption = AbstractOption
