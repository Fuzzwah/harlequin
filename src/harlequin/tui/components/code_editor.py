from textual.binding import Binding
from textual.widgets import Input


class CodeEditor(Input):
    BINDINGS = [
        ("ctrl+enter", "submit", "Run Query"),
        Binding("ctrl+j", "submit", "Run Query", show=False),
        Binding("enter", "newline", "", show=False),
    ]

    def on_mount(self) -> None:
        self.border_title = "Code Editor"

    def action_newline(self) -> None:
        self.insert_text_at_cursor("\n")
