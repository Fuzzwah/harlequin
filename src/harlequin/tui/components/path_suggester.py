from pathlib import Path
from typing import Union

from textual.suggester import Suggester


class PathSuggester(Suggester):
    def __init__(self) -> None:
        super().__init__(use_cache=True, case_sensitive=True)

    async def get_suggestion(self, value: str) -> Union[str, None]:
        try:
            p = Path(value).expanduser()
            matches = list(p.parent.glob(f"{p.parts[-1]}*"))
            if len(matches) == 1:
                return str(matches[0])
            else:
                return None
        except Exception:
            return None
