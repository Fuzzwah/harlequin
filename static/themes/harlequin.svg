version https://git-lfs.github.com/spec/v1
oid sha256:7aac3e860404ee956b1ee38dc75337ac8f1d470b5975c7e0e8dc491e8dd4089e
size 143134
