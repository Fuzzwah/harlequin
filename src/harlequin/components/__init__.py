from harlequin.components.code_editor import CodeEditor, EditorCollection
from harlequin.components.error_modal import ErrorModal
from harlequin.components.export_screen import (
    CSVOptions,
    ExportOptions,
    ExportScreen,
    JSONOptions,
    ParquetOptions,
)
from harlequin.components.help_screen import HelpScreen
from harlequin.components.results_viewer import ResultsViewer
from harlequin.components.run_query_bar import RunQueryBar
from harlequin.components.schema_viewer import CatalogItem, SchemaViewer

__all__ = [
    "CodeEditor",
    "EditorCollection",
    "ErrorModal",
    "ExportOptions",
    "CSVOptions",
    "ParquetOptions",
    "JSONOptions",
    "ExportScreen",
    "HelpScreen",
    "ResultsViewer",
    "RunQueryBar",
    "CatalogItem",
    "SchemaViewer",
]
