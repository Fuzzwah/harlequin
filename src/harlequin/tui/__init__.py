from harlequin.tui.code_editor import CodeEditor as CodeEditor
from harlequin.tui.results_viewer import ResultsViewer as ResultsViewer
from harlequin.tui.schema_viewer import (
    SCHEMAS as SCHEMAS,
)
from harlequin.tui.schema_viewer import (
    TABLES as TABLES,
)
from harlequin.tui.schema_viewer import (
    SchemaViewer as SchemaViewer,
)
from harlequin.tui.error_modal import ErrorModal as ErrorModal
