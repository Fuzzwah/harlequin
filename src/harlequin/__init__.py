from harlequin.tui.sql_client import Harlequin as Harlequin
