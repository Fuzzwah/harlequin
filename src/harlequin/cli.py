from pathlib import Path
import click
from harlequin import Harlequin
import duckdb


@click.command()
@click.version_option(package_name="harlequin")
@click.argument(
    "db_path",
    default=":memory:",
    nargs=1,
    type=click.Path(path_type=Path),
)
def harlequin(db_path: Path) -> None:
    conn = duckdb.connect(str(db_path))
    tui = Harlequin(connection=conn)
    tui.run()
