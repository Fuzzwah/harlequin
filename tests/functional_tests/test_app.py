import pytest
from harlequin.tui import Harlequin
from harlequin.tui.components import CodeEditor


@pytest.mark.asyncio
async def test_select_1(app: Harlequin) -> None:
    async with app.run_test() as pilot:
        assert app.title == "Harlequin"
        assert app.focused.__class__.__name__ == "TextInput"

        q = "select 1 as foo"
        for key in q:
            await pilot.press(key)
        await pilot.press("ctrl+j")  # alias for ctrl+enter

        # when the query is submitted, it should update app.query_text, app.relation,
        # and app.data using three different workers.
        await app.workers.wait_for_complete()
        assert app.query_text == q
        await app.workers.wait_for_complete()
        assert app.relation is not None
        await app.workers.wait_for_complete()
        assert app.data == [(1,)]


@pytest.mark.asyncio
async def test_query_formatting(app: Harlequin) -> None:
    async with app.run_test() as pilot:
        editor = app.query_one(CodeEditor)
        editor.text = "select\n\n1 FROM\n\n foo"

        await pilot.press("ctrl+@")  # alias for ctrl+`
        assert editor.text == "select 1 from foo\n"
