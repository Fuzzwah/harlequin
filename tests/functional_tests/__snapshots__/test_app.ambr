version https://git-lfs.github.com/spec/v1
oid sha256:98406c9a7e26056f2f4435c63c6814cb80cf18c1f99e13e6a8cbb61aab6e9d99
size 3774648
